custom_components/localtuya/coordinator.py

```diff
@@ -428,40 +437,47 @@ async def set_dps(self, states):
     async def _async_refresh(self, _now):
         if self.connected:
             self.debug("Refreshing dps for device")
-            # This a workdaround for >= 3.4 devices, since there is an issue on waiting for the correct seqno
+            # This a workaround for >= 3.4 devices, since there is an issue on waiting for the correct seqno
             try:
                 await self._interface.update_dps(cid=self._node_id)
             except TimeoutError:
                 pass
 
+    async def _sleep(self, seconds) -> bool:
```

Comment:
Now that we handled the cancel issues on close with ignoring cancel error, this is not needed.

Reply:
At least `_task_shutdown_entities` is cancelled also from `disconnected()`. Without this `_sleep()`, it is not calcelled and continue living for `self._device_config.sleep_time` seconds (very long).

Well, currently the required `try` block can be implemented inside `_task_reconnect` itself, but I believe the common fail proof solution is better.

Reply:
It does canceled fine from `disconnected` I have tested as I mentioned because we aren't awaiting for it clean up it's fine the issue was in before close because we were waiting for them to finish without try block which leads rising an error without handling it.

Reply:
> It does canceled fine from disconnected I have tested

How did you test? I mean, how do you check that after `self._task_shutdown_entities.cancel()` call the task indeed ends and does not continue sleeping?

Reply:
I re-ran my tests without the `_sleep()` and I confirm it is not required. Before, I noticed hungs in the close on `await self._task_shutdown_entities`, now it is gone. Probably, because of your `await asyncio.gather(*pending_tasks, return_exceptions=True)` change.

I'm yet to check the other changes you suggested in the https://github.com/xZetsubou/hass-localtuya/pull/383#discussion_r1844282901... but I feel we are close to the end 😄